### PR TITLE
Addition of SPEEDY

### PR DESCRIPTION
**Pull request by DanGiles**

### CODE REVIEW

**Comment by tkoskela** [comment]
Hi @DanGiles is this ok for us to review or do you want to keep working on it since it's a draft?

**Comment by DanGiles** [comment]
> Hi @DanGiles is this ok for us to review or do you want to keep working on it since it's a draft?

This is now ready for review. However, please note that there are some outstanding issues.

> The current optimal proposal implementation does not work for the SPEEDY model. The optimal filter will run but is not correct.

> The performance can be improved by removing the unnecessary memory allocation for the 2D variables.

> Details on the compilation steps for the SPEEDY model need to be added.

File: speedy/model/model.jl [hunk 1]
```diff
@@ -0,0 +1,855 @@
...
+    # Bash script call to speedy
+    run(`$forecast $SPEEDY $output $YMDH $TYMDH $rank $particle`)
+end
+
+function ParticleDA.update_particle_dynamics!(d::ModelData, nprt_per_rank)
+    my_rank = MPI.Comm_rank(MPI.COMM_WORLD)
+
+    Threads.@threads for ip in 1:nprt_per_rank
+        #Write to file
+        anal_file = string(d.model_params.anal_folder,my_rank,"/",ip,"/",d.dates[1],".grd")
```
Comment (giordano) [inline speedy/model/model.jl:616]:
```suggestion
        anal_file = joinpath(d.model_params.anal_folder, my_rank, ip,d.dates[1] * ".grd")
```

File: speedy/model/model.jl [hunk 2]
```diff
@@ -0,0 +1,855 @@
...
+        write_fortran(anal_file,d.model_params.nlon, d.model_params.nlat, d.model_params.nlev,d.states.particles[:, :, :, :, ip])
+        # Update the dynamics
+        speedy_update!(d.model_params.SPEEDY,d.model_params.output_folder,d.dates[1],d.dates[2],string(my_rank),string(ip))
+        # Read back in the data and update the states
+        guess_file = string(d.model_params.guess_folder,my_rank,"/",ip,"/",d.dates[2],".grd")
```
Comment (giordano) [inline speedy/model/model.jl:621]:
```suggestion
        guess_file = joinpath(d.model_params.guess_folder, my_rank, ip, d.dates[2] * ".grd")
```

File: src/ParticleDA.jl [hunk 3]
```diff
@@ -361,7 +368,7 @@ function init_filter(filter_params::FilterParameters, model_data, nprt_per_rank:
     domain_size = get_grid_domain_size(model_data)
     cell_size = get_grid_cell_size(model_data)
 
-    grid = Grid(size...,cell_size...,domain_size...)
+    grid = Grid(size[1],size[2],cell_size[1],cell_size[2],domain_size[1],domain_size[2])
```
Comment (giordano) [inline src/ParticleDA.jl:371]: This is assuming again that `size`, `cell_size` and `domain_size` have specific dimensions, isn't it?

File: speedy/extracting_speedy.jl [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+nlev= 8
+n_state_var= 6
+truth = zeros(T, nlon, nlat, nlev, n_state_var);
+output_filename = "DATA/nature_runs.h5"
+
+iteration = 0
+for file in readdir(nature_dir)
+    if iteration < 242
+        if occursin(".grd", file)
+            @show joinpath(nature_dir,file)
```
Comment (giordano) [inline speedy/extracting_speedy.jl:119]: `@show` is useful for quick debugging, but it shouldn't appear in production code.

File: speedy/extracting_speedy.jl [hunk 5]
```diff
@@ -0,0 +1,126 @@
+using FortranFiles
```
Comment (giordano) [inline speedy/extracting_speedy.jl:1]: Tuomas pointed out this file may not actually be used anywhere.

**Review by giordano — COMMENTED** [review]
First round of quick comments

File: speedy/extracting_speedy.jl [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        attributes(ds)["Description"] = description
+        attributes(ds)["Unit"] = unit
+        attributes(ds)["Time step"] = it
+        # attributes(ds)["Date"] = d.dates[1]
+    else
+        @warn "Write failed, dataset " * group_name * "/" * subgroup_name * "/" * dataset_name *  " already exists in " * file.filename * "!"
+    end
+end
+
+SPEEDY= "/Users/dangiles/Documents/UCL/Raddish/speedy"
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/extracting_speedy.jl:105]: My bad..

```suggestion
SPEEDY= ""
```

File: speedy/extracting_speedy.jl [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+nature_dir = string(SPEEDY,"/DATA/nature/")
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/extracting_speedy.jl:106]:
```suggestion
nature_dir = joinpath(SPEEDY, "DATA", "nature")
```

File: speedy/extracting_speedy.jl (hunk 4 above)
Comment (DanGiles) [inline speedy/extracting_speedy.jl:119]:
```suggestion
```

File: speedy/extracting_speedy.jl [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    if iteration < 242
+        if occursin(".grd", file)
+            @show joinpath(nature_dir,file)
+            read_grd!(joinpath(nature_dir,file), nlon, nlat, nlev, @view(truth[:,:,:,:]))
+            write_snapshot(output_filename, truth, iteration)
+            # @show truth[:,:,1,5]
+            iteration = iteration + 1
+        end
+    end 
+end
```
Comment (DanGiles) [inline speedy/extracting_speedy.jl:126]:
```suggestion
end

```

File: speedy/model/model.jl [hunk 9]
```diff
@@ -0,0 +1,855 @@
...
+    particle_dump_time = [-1]
+    obs_noise_std::T = 1000.0
+    #Path to the the local speedy directory
+    SPEEDY::String = "/Users/dangiles/Documents/UCL/Raddish/speedy"
+    station_filename::String = string(SPEEDY,"/obs/networks/",obs_network,".txt")
+    nature_dir::String = string(SPEEDY,"/DATA/nature/")
+    # Output folders
+    output_folder::String = string(pwd(),"/speedy")
+    guess_folder::String = string(output_folder,"/DATA/ensemble/gues/")
+    anal_folder::String = string(output_folder,"/DATA/ensemble/anal/")
```
Comment (DanGiles) [inline speedy/model/model.jl:98]:
```suggestion
    SPEEDY::String = ""
    station_filename::String = joinpath(SPEEDY, "obs", "networks", "obs_network" * ".txt")
    nature_dir::String = joinpath(SPEEDY, "DATA", "nature")
    # Output folders
    output_folder::String = joinpath(pwd(), "speedy")
    guess_folder::String = joinpath(output_folder, "DATA", "ensemble", "gues")
    anal_folder::String = joinpath(output_folder, "DATA", "ensemble", "anal")
```

File: speedy/model/model.jl [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+ParticleDA.get_grid_domain_size(d::ModelData) = d.model_params.lon_length,d.model_params.lat_length,d.model_params.z_length
+ParticleDA.get_grid_cell_size(d::ModelData) = d.model_params.dx,d.model_params.dy,d.model_params.dz
+ParticleDA.get_n_state_var(d::ModelData) = d.model_params.n_state_var
+
+
+function create_folders(output_folder::String, anal_folder::String, gues_folder::String, nprt_per_rank::Int, my_rank::Integer)
+    MPI.Init()
+    comm = MPI.COMM_WORLD
+    ens = string(output_folder,"/DATA/ensemble/")
+    tmp = string(output_folder,"/DATA/tmp/ensfcst/")
```
Comment (DanGiles) [inline speedy/model/model.jl:463]:
```suggestion
    ens = joinpath(output_folder, "DATA", "ensemble")
    tmp = joinpath(output_folder, "DATA", "tmp", "ensfcst")
```

File: speedy/model/model.jl [hunk 11]
```diff
@@ -0,0 +1,855 @@
...
+        run(`mkdir $ens`)
+        run(`mkdir $tmp`)
+        run(`mkdir $anal_folder`)
+        run(`mkdir $gues_folder`)
+    end
+    MPI.Barrier(comm)
+
+    rank_anal = string(anal_folder,my_rank)
+    rank_gues = string(gues_folder,my_rank)
+    rank_tmp = string(tmp,my_rank)
```
Comment (DanGiles) [inline speedy/model/model.jl:476]:
```suggestion
    rank_anal = joinpath(anal_folder,my_rank)
    rank_gues = joinpath(gues_folder,my_rank)
    rank_tmp = joinpath(tmp,my_rank)
```

File: speedy/model/model.jl [hunk 12]
```diff
@@ -0,0 +1,855 @@
...
+    run(`mkdir $rank_tmp`)
+    run(`mkdir $rank_anal`)
+    run(`mkdir $rank_gues`)
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:479]:
```suggestion
    mkdir(rank_tmp)
    mkdir(rank_anal)
    mkdir(rank_gues)
```

File: speedy/model/model.jl [hunk 13]
```diff
@@ -0,0 +1,855 @@
...
+    Threads.@threads for ip in 1:nprt_per_rank
+        part_anal = string(rank_anal,"/",ip)
+        part_gues = string(rank_gues,"/",ip)
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:482]:
```suggestion
        part_anal = joinpath(rank_anal, ip)
        part_gues = joinpath(rank_gues, ip)
```

File: speedy/model/model.jl [hunk 14]
```diff
@@ -0,0 +1,855 @@
...
+        run(`mkdir $part_anal`)
+        run(`mkdir $part_gues`)
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:484]:
```suggestion
        mkdir(part_anal)
        mkdir(part_gues)
```

File: speedy/model/model.jl [hunk 15]
```diff
@@ -0,0 +1,855 @@
...
+    for n = 1:nv2d
+        truth[:,:,1,(n+nv3d)] .= read(f, (Float32, nlon, nlat), rec = irec)
+        irec += 1
+    end
+
+    close(f)
+
+end
+
+function read_ps!(filename::String,nlon::Int, nlat::Int, nlev::Int,truth::AbstractMatrix{T}) where T
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:533]:
```suggestion
function read_ps!(truth::AbstractMatrix{T}, filename::String, nlon::Int, nlat::Int, nlev::Int) where T
```

File: speedy/model/model.jl [hunk 16]
```diff
@@ -0,0 +1,855 @@
...
+                            nprt_per_rank::Int,
+                            params::ModelParameters) where T
+
+    # Set true initial state
+    # Read in the initial nature run - just surface pressure
+    read_ps!(string(params.nature_dir,params.IDate,".grd"), params.nlon, params.nlat, params.nlev, @view(states.truth[:,:,1]))
+    ### Read in arbitrary nature run files for the initial conditions
+    Threads.@threads for ip in 1:nprt_per_rank
+        dummy_date = step_ens(params.ensDate, params.Hinc)
+        read_grd!(string(params.nature_dir,dummy_date,".grd"), params.nlon, params.nlat, params.nlev, @view(states.particles[:,:,:,:,ip]))
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:378]: Due to altered definition of the function:

```suggestion
        read_grd!(@view(states.particles[:,:,:,:,ip]), joinpath(params.nature_dir, dummy_date * ".grd"), params.nlon, params.nlat, params.nlev, @view(states.particles[:,:,:,:,ip]))
```

File: speedy/model/model.jl [hunk 17]
```diff
@@ -0,0 +1,855 @@
...
+    for n = 1:nv2d
+        write(f,v2d[:,:,n],rec=irec)
+        irec=irec+1
+    end
+    close(f)
+end
+
+function ParticleDA.update_truth!(d::ModelData, _)
+    # Read in Nature run
+    read_ps!(string(d.model_params.nature_dir,d.dates[1],".grd"), d.model_params.nlon, d.model_params.nlat, d.model_params.nlev, @view(d.states.truth[:,:,1]))
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:592]: Yes and with the altered definition of read_ps!()
```suggestion
    read_ps!(@view(d.states.truth[:,:,1]), d.model_params.nlon, d.model_params.nlat, d.model_params.nlev, joinpath(d.model_params.nature_dir, d.dates[1] * ".grd"))
```

File: src/ParticleDA.jl (hunk 3 above)
Comment (DanGiles) [reply src/ParticleDA.jl:371]: If we try to pass three dimensions to init_offline_matrices() we still get an error. Therefore, I just pass the two horizontal components

File: speedy/extracting_speedy.jl (hunk 5 above)
Comment (DanGiles) [reply speedy/extracting_speedy.jl:1]: Yes this is correct, I can remove this. I was using it to extract the speedy outputs and then saving them to a .h5 file

File: speedy/extracting_speedy.jl [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+
+SPEEDY= "/Users/dangiles/Documents/UCL/Raddish/speedy"
+nature_dir = string(SPEEDY,"/DATA/nature/")
+T = Float64
+nlon = 96
+nlat= 48
+nlev= 8
+n_state_var= 6
+truth = zeros(T, nlon, nlat, nlev, n_state_var);
+output_filename = "DATA/nature_runs.h5"
```
Comment (DanGiles) [inline speedy/extracting_speedy.jl:113]:
```suggestion
output_filename = joinpath("DATA", "nature_runs.h5")
```

File: speedy/model/dafcst.sh [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+ln -fs $OUTPUT/gues/${rank}/${particle}/fluxes.grd fluxes.grd
+FORT2=2
+echo $FORT2 > fort.2
+echo $YMDH | cut -c1-4 >> fort.2
+echo $YMDH | cut -c5-6 >> fort.2
+echo $YMDH | cut -c7-8 >> fort.2
+echo $YMDH | cut -c9-10 >> fort.2
+./imp > out.lis 2> out.lis.2
+mv ${TYMDH}.grd $OUTPUT/gues/${rank}/${particle}
+echo "Finished"
```
Comment (DanGiles) [inline speedy/model/dafcst.sh:50]:
```suggestion
```

File: speedy/model/model.jl [hunk 20]
```diff
@@ -0,0 +1,855 @@
...
+
+    lambda::Vector{T} = [1.0e0, 1.0e0, 1.0e0, 1.0e0, 1.0e0]
+    nu::Vector{T} = [2.5, 2.5, 2.5, 2.5, 2.5]
+    sigma::Vector{T} = [0.1, 0.1, 0.1, 0.00001, 1000.0]
+
+    lambda_initial_state::Vector{T} = [1.0e0, 1.0e0, 1.0e0, 1.0e0, 1.0e0]
+    nu_initial_state::Vector{T} = [2.5, 2.5, 2.5, 2.5, 2.5]
+    sigma_initial_state::Vector{T} = [0.1, 0.1, 0.1, 0.0001, 1000.0]
+
+    padding::Int = 100
```
Comment (DanGiles) [inline speedy/model/model.jl:78]:
```suggestion
```

File: speedy/model/model.jl [hunk 21]
```diff
@@ -0,0 +1,855 @@
...
+    readline(f)
+    readline(f)
+    ind = 0
+    for line in eachline(f)
+        if ind%10 == 0
+            append!(ist, parse(Int64,split(line)[1]))
+            append!(jst, parse(Int64,split(line)[2]))
+        end
+        ind = ind + 1
+    end
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:420]: The files are simple two column text files with each row giving separate locations

File: speedy/model/model.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:476]:
```suggestion
    rank_anal = joinpath(anal_folder,my_rank)
    rank_gues = joinpath(gues_folder,my_rank)
    rank_tmp = joinpath(tmp,my_rank)
```

File: speedy/model/model.jl [hunk 22]
```diff
@@ -0,0 +1,855 @@
...
+    model_dates = [model_params.IDate,model_params.dtDate]
+    set_stations!(stations, model_params)
+
+    set_initial_state!(states, model_matrices, field_buffer, rng, nprt_per_rank, model_params)
+
+    create_folders(model_params.output_folder, model_params.anal_folder, model_params.guess_folder, nprt_per_rank, my_rank)
+
+    return ModelData(model_params, states, observations, stations, field_buffer, background_grf, model_matrices, rng, model_dates)
+end
+function read_grd!(filename::String,nlon::Int, nlat::Int, nlev::Int,truth::AbstractArray{T}) where T
```
Comment (DanGiles) [inline speedy/model/model.jl:503]:
```suggestion
function read_grd!(truth::AbstractArray{T}, filename::String,nlon::Int, nlat::Int, nlev::Int) where T
```

File: speedy/model/model.jl [hunk 23]
```diff
@@ -0,0 +1,855 @@
...
+
+    Threads.@threads for ip in 1:nprt_per_rank
+        #Write to file
+        anal_file = string(d.model_params.anal_folder,my_rank,"/",ip,"/",d.dates[1],".grd")
+        write_fortran(anal_file,d.model_params.nlon, d.model_params.nlat, d.model_params.nlev,d.states.particles[:, :, :, :, ip])
+        # Update the dynamics
+        speedy_update!(d.model_params.SPEEDY,d.model_params.output_folder,d.dates[1],d.dates[2],string(my_rank),string(ip))
+        # Read back in the data and update the states
+        guess_file = string(d.model_params.guess_folder,my_rank,"/",ip,"/",d.dates[2],".grd")
+        read_grd!(guess_file, d.model_params.nlon, d.model_params.nlat, d.model_params.nlev,@view(d.states.particles[:, :, :, :, ip]))
```
Comment (DanGiles) [inline speedy/model/model.jl:622]:
```suggestion
        read_grd!(@view(d.states.particles[:, :, :, :, ip]), guess_file, d.model_params.nlon, d.model_params.nlat, d.model_params.nlev)
```

File: speedy/model/dafcst.sh [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+
+# Input for this shell
+SPEEDY=$1
+OUTPUT=$2
+YMDH=$3
+TYMDH=$4
+rank=$5
+particle=$6
+
+TMPDIR=$2/DATA/tmp
```
Comment (DanGiles) [inline speedy/model/dafcst.sh:15]: Should all these paths also be altered? If run on a windows?

File: speedy/model/model.jl [hunk 25]
```diff
@@ -0,0 +1,855 @@
...
+end
+
+function speedy_update!(SPEEDY::String,
+                        output::String,
+                        YMDH::String,
+                        TYMDH::String,
+                        rank::String,
+                        particle::String)
+    # Path to the bash script which carries out the forecast
+    forecast = "./speedy/model/dafcst.sh"
```
Comment (DanGiles) [inline speedy/model/model.jl:606]:
```suggestion
    forecast = joinpath("speedy", "model", "dafcst.sh")
```

File: speedy/model/model.jl [hunk 26]
```diff
@@ -0,0 +1,855 @@
...
+#         title=attr(
+#         text= string("Model Noise for Surface Pressure"),
+#         xanchor= "center",
+#         yanchor= "top"),
+#         xaxis_title="Lon (•)",
+#         yaxis_title="Lat (•)",
+#         )
+#     ps = plot(trace, layout)
+#     savefig(ps,string("/Users/dangiles/.julia/dev/ParticleDA/results/speedy/",string(i),".png"))
+# end
```
Comment (DanGiles) [inline speedy/model/model.jl:363]: Only used for debugging
```suggestion
```

File: speedy/model/model.jl [hunk 27]
```diff
@@ -0,0 +1,820 @@
...
+    new_idate = Dates.format(DateTime(idate, "YYYYmmddHH") + Dates.Hour(6),"YYYYmmddHH")
+    new_dtdate = Dates.format(DateTime(dtdate, "YYYYmmddHH") + Dates.Hour(6),"YYYYmmddHH")
+    return new_idate,new_dtdate
+end
+
+function step_ens(ensdate::String, Hinc::Int)
+    ens_end = Dates.format(DateTime(ensdate, "YYYYmmddHH") + Dates.Month(1),"YYYYmmddHH")
+    diff = length(DateTime(ensdate, "YYYYmmddHH"):Hour(Hinc):DateTime(ens_end, "YYYYmmddHH"))
+    rand_int = rand(0:diff-1)*Hinc
+    new_idate = Dates.format(DateTime(ensdate, "YYYYmmddHH") + Dates.Hour(rand_int),"YYYYmmddHH")
```
Comment (giordano) [inline speedy/model/model.jl:128]: If you don't want to repeat `YYYYmmddHH` everywhere, you can do something like
```suggestion

const SPEEDY_DATE_FORMAT = "YYYYmmddHH"

function step_datetime(idate::String,dtdate::String)
    new_idate = Dates.format(DateTime(idate, SPEEDY_DATE_FORMAT) + Dates.Hour(6), SPEEDY_DATE_FORMAT)
    new_dtdate = Dates.format(DateTime(dtdate, SPEEDY_DATE_FORMAT) + Dates.Hour(6), SPEEDY_DATE_FORMAT)
    return new_idate,new_dtdate
end

function step_ens(ensdate::String, Hinc::Int)
    ens_end = Dates.format(DateTime(ensdate, SPEEDY_DATE_FORMAT) + Dates.Month(1), SPEEDY_DATE_FORMAT)
    diff = length(DateTime(ensdate, SPEEDY_DATE_FORMAT):Hour(Hinc):DateTime(ens_end, SPEEDY_DATE_FORMAT))
    rand_int = rand(0:diff-1)*Hinc
    new_idate = Dates.format(DateTime(ensdate, SPEEDY_DATE_FORMAT) + Dates.Hour(rand_int), SPEEDY_DATE_FORMAT)
```

File: speedy/model/model.jl [hunk 28]
```diff
@@ -0,0 +1,820 @@
...
+    return [_generate(l, n, s) for (l, n, s) in zip(lambda, nu, sigma)]
+end
+# Get a random sample from random_field_generator using random number generator rng
+function sample_gaussian_random_field!(field::AbstractArray{T,2},
+                                       random_field_generator::RandomField,
+                                       rng::Random.AbstractRNG) where T
+
+    field .= GaussianRandomFields.sample(random_field_generator.grf)
+    # @. @view(random_field_generator.xi[:,:,threadid()]) = randn((rng,), T)
+    # sample_gaussian_random_field!(field, random_field_generator, @view(random_field_generator.xi[:,:,threadid()]))
```
Comment (giordano) [inline speedy/model/model.jl:254]: It'd be good to remove all the commented out code :wink:

File: speedy/model/model.jl [hunk 29]
```diff
@@ -0,0 +1,820 @@
...
+using ParticleDA
+
+using Random, Distributions, Base.Threads, GaussianRandomFields, HDF5
+# using Default_params
+using DelimitedFiles
+using FieldMetadata
+using FortranFiles
+using Setfield
+using Dates
+using PlotlyJS
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply speedy/model/model.jl:12]:
```suggestion
```

File: speedy/model/model.jl [hunk 30]
```diff
@@ -0,0 +1,820 @@
...
+#                                        random_field_generator::RandomField,
+#                                        random_numbers::AbstractArray{T}) where T
+
+#     field .= GaussianRandomFields._sample!(@view(random_field_generator.w[:,:,threadid()]),
+#                                            @view(random_field_generator.z[:,:,threadid()]),
+#                                            random_field_generator.grf,
+#                                            random_numbers)
+
+# end
+
```
Comment (DanGiles) [inline speedy/model/model.jl:269]:
```suggestion
```

File: speedy/model/model.jl (hunk 28 above)
Comment (DanGiles) [reply speedy/model/model.jl:254]:
```suggestion
```

File: src/ParticleDA.jl [hunk 31]
```diff
@@ -42,6 +42,13 @@ Return the number of state variables.
 """
 function get_n_state_var end
 
+"""
+    ParticleDA.get_indices(model_data) -> Vector{Int}
+
+Return the vector containing the indices of assimilated values in the state vector
+"""
+function get_indices end
```
[the remark this answers is not in the document]
Comment (DanGiles) [reply src/ParticleDA.jl:50]: A tuple might not work if at some point one wanted to assimilate more than one variable (and at different elevations)?

**Comment by matt-graham** [comment]
I believe the SPEEDY model code and continued development has now been transferred by @DanGiles to the separate [Team-RADDISH/ParticleDA-UseCases](https://github.com/Team-RADDISH/ParticleDA-UseCases) repository and so this can now be closed?

**Pull request closed without merging**